File: AdventOfCode2021/d12/d12.dart
import 'dart:io';

List<List<String>> readInp() {
    List<List<String>> out = [];
    while(true) {
        var line = stdin.readLineSync();
        if(line == null)
            break;
        out.add(line.split("-"));
    }
    return out;
}

typedef Graph = Map<String, List<String>>;
void push(Graph mp, String k, String v) {
    mp.update(k, (cur) { return cur + [v]; }, ifAbsent: (){return [v];});
}

Graph buildGraph(List<List<String>> inp) {
    Graph mp = Map();
    for(final line in inp) {
        var a = line[0];
        var b = line[1];
        push(mp, a, b);
        push(mp, b, a);
    }
    return mp;
}

bool isSmall(String wrd) {
    for(int i = 0; i < wrd.length; i++) {
        var ch = wrd[i];
        if(ch.compareTo('a') < 0 || ch.compareTo('z') > 0)
            return false;
    }
    return true;
}

int explore(Graph g, String c, Set<String> vis, bool usedDV) {
    if(c == "end")
        return 1;
    var sm = 0;
    for(final n in g[c]!) {
        if(!(isSmall(n) && vis.contains(n))) {
            if(isSmall(n))
                vis.add(n);
            sm += explore(g, n, vis, usedDV);
            if(isSmall(n))
                vis.remove(n);
        }
        if(isSmall(n) && vis.contains(n) && !usedDV && n != "start")
            sm += explore(g, n, vis, true);
    }
    return sm;
}

void partA(Graph g) {
    print(explore(g, "start", Set.from(["start"]), true));
}

void partB(Graph g) {
    print(explore(g, "start", Set.from(["start"]), false));
}

void main() {
    var inp = readInp();
    var g = buildGraph(inp);
    partA(g);
    partB(g);
}

File: AdventOfCode2021/pysolns/d12.py
import sys
from collections import defaultdict

def lines():
  return [line.strip() for line in sys.stdin]

inp = [tuple(l.split('-')) for l in lines()]
grph = defaultdict(list)
for a,b in inp:
  grph[a].append(b)
  grph[b].append(a)

def small(cv):
  return all('a' <= x <= 'z' for x in cv)

def explore(cv, vis, usedDV):
  if cv == "end":
    return 1
  sm = 0
  for n in grph[cv]:
    if not (small(n) and n in vis):
      if small(n):
        vis.add(n)
      sm += explore(n, vis, usedDV)
      if small(n):
        vis.remove(n)
    if small(n) and n in vis and not usedDV and n != "start":
      sm += explore(n, vis, True)

  return sm

def partA():
  print(explore("start", set(["start"]), True))

def partB():
  print(explore("start", set(["start"]), False))
  pass

partA()
partB()
